File: backend/src/main/java/com/matrimonial/service/InterestService.java
package com.matrimonial.service;

import com.matrimonial.entity.Interest;
import com.matrimonial.entity.User;
import com.matrimonial.exception.InterestNotFoundException;
import com.matrimonial.repository.InterestRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class InterestService {

    private static final String PENDING = "PENDING";
    private static final String ACCEPTED = "ACCEPTED";
    private static final String REJECTED = "REJECTED";

    private final InterestRepository interestRepository;
    private final UserService userService;

    public InterestService(InterestRepository interestRepository, UserService userService) {
        this.interestRepository = interestRepository;
        this.userService = userService;
    }

    // FIX ISSUE-20: @Transactional prevents race condition between the duplicate
    // existence check and the save — concurrent duplicate sends are blocked at DB level.
    @Transactional
    public Interest sendInterest(Long senderId, Long receiverId) {
        if (senderId.equals(receiverId)) {
            throw new IllegalArgumentException("A user cannot send interest to themselves");
        }
        if (interestRepository.existsBySenderUserIdAndReceiverUserIdAndStatus(senderId, receiverId, PENDING)) {
            throw new IllegalStateException("An interest request is already pending for this user");
        }
        User sender = userService.getUserById(senderId);
        User receiver = userService.getUserById(receiverId);

        Interest interest = new Interest();
        interest.setSender(sender);
        interest.setReceiver(receiver);
        interest.setStatus(PENDING);
        return interestRepository.save(interest);
    }

    public List<Interest> getSentInterests(Long senderId) {
        return interestRepository.findBySenderUserId(senderId);
    }

    public List<Interest> getReceivedInterests(Long receiverId) {
        return interestRepository.findByReceiverUserId(receiverId);
    }

    @Transactional
    public Interest acceptInterest(Long interestId) {
        return updateStatus(interestId, ACCEPTED);
    }

    @Transactional
    public Interest rejectInterest(Long interestId) {
        return updateStatus(interestId, REJECTED);
    }

    @Transactional
    public void deleteInterest(Long interestId) {
        Interest interest = getInterestById(interestId);
        interestRepository.delete(interest);
    }

    /**
     * A match exists when both users have ACCEPTED interests toward each other.
     * One-directional accept is not enough — chat unlocks only on mutual match.
     */
    public boolean isMatched(Long user1Id, Long user2Id) {
        return interestRepository.existsBySenderUserIdAndReceiverUserIdAndStatus(user1Id, user2Id, ACCEPTED)
                && interestRepository.existsBySenderUserIdAndReceiverUserIdAndStatus(user2Id, user1Id, ACCEPTED);
    }

    private Interest updateStatus(Long interestId, String status) {
        Interest interest = getInterestById(interestId);
        interest.setStatus(status);
        return interestRepository.save(interest);
    }

    private Interest getInterestById(Long interestId) {
        return interestRepository.findById(interestId)
                .orElseThrow(() -> new InterestNotFoundException("Interest not found with ID: " + interestId));
    }
}

File: backend-flask/app/services/interest_service.py
"""Port of InterestService. Same rules, same messages."""
from ..errors import BadRequestError, ConflictError, NotFoundError
from ..extensions import db
from ..models import Interest
from .profile_service import get_user_or_404

PENDING = "PENDING"
ACCEPTED = "ACCEPTED"
REJECTED = "REJECTED"


def send_interest(sender_id, receiver_id):
    if sender_id == receiver_id:
        raise BadRequestError("A user cannot send interest to themselves")
    exists = Interest.query.filter_by(
        sender_id=sender_id, receiver_id=receiver_id, status=PENDING).first()
    if exists is not None:
        raise ConflictError("An interest request is already pending for this user")
    sender = get_user_or_404(sender_id)
    receiver = get_user_or_404(receiver_id)

    interest = Interest(sender_id=sender.user_id, receiver_id=receiver.user_id,
                        status=PENDING)
    db.session.add(interest)
    db.session.commit()
    return interest


def get_sent_interests(sender_id):
    return Interest.query.filter_by(sender_id=sender_id).all()


def get_received_interests(receiver_id):
    return Interest.query.filter_by(receiver_id=receiver_id).all()


def _get_or_404(interest_id):
    interest = db.session.get(Interest, interest_id)
    if interest is None:
        raise NotFoundError(f"Interest not found with ID: {interest_id}")
    return interest


def accept_interest(interest_id):
    interest = _get_or_404(interest_id)
    interest.status = ACCEPTED
    db.session.commit()
    return interest


def reject_interest(interest_id):
    interest = _get_or_404(interest_id)
    interest.status = REJECTED
    db.session.commit()
    return interest


def delete_interest(interest_id):
    interest = _get_or_404(interest_id)
    db.session.delete(interest)
    db.session.commit()


def is_matched(user1_id, user2_id):
    """A match exists when both users have ACCEPTED interests toward each other."""
    forward = Interest.query.filter_by(
        sender_id=user1_id, receiver_id=user2_id, status=ACCEPTED).first()
    backward = Interest.query.filter_by(
        sender_id=user2_id, receiver_id=user1_id, status=ACCEPTED).first()
    return forward is not None and backward is not None
